### PR TITLE
Final ASMAPI cleanup, NumberType enum change, and new insertInsnList method

### PR DESCRIPTION
**Pull request by Jonathing**
This PR is, what I hope to be, my final cleanup PR for ASMAPI. There were a lot of things with it I wanted to change so that it could see better usage in CoreMods without needing to resort to (too much) extra work on the modder's part, given that CoreMods is sandboxed.

- Renamed `appendMethodCall` to `injectMethodCall`.
  - The former still exists for compatibility.
- Renamed `LdcNumberType` to `NumberType`.
- Added `castNumber`, a method which returns an Object of the desired value casted to the given NumberType.
  - I realized that there are many cases where a direct object value (usually the primitive types) need to be fed into an instruction directly, the best example being `FieldNode`. This approach allows modders to cast numbers themselves without needing to rely on us to make additional stray methods into ASMAPI.
- Added a new `insertInsnList` method which accepts `AbstractInsnNode` as opposed to the components of a `MethodInsnNode`.
  - This allows for modders to use ASMAPI to insert their instruction lists using ASMAPI's `InsertMode`, reducing the need to interface with the method node's instruction list directly.
  - As an additional code cleanup, the old `insertInsnList` now uses this new method along with `findFirstMethodCall`, which is essentially what it was doing in the first place.

My goal with these ASMAPI changes is to have them be added into Forge builds as far back as 1.18.2. These old versions only have so much to gain from these changes. The only logical breaking change is the bug fix to `findFirstInstructionBefore` which was added in #50.

Once this is merged, we can look into enhancing the Nashorn engine that CoreMods uses [as described in this wishlist.](https://gist.github.com/PaintNinja/4d1a35c03b377e78a74c708184ce893e#coremods)

### CODE REVIEW

File: src/main/java/net/minecraftforge/coremod/api/ASMAPI.java [hunk 1]
```diff
@@ -33,17 +32,29 @@ public static MethodNode getMethodNode() {
...
 
+    /**
+     * Injects a method call to the beginning of the given method.
+     *
+     * @param node       The method to inject the call into
+     * @param methodCall The method call to inject
+     *
+     * @deprecated Renamed to {@link #injectMethodCall(MethodNode, MethodInsnNode)}
+     */
+    @Deprecated(forRemoval = true, since = "5.1")
```
Comment (Jonathing) [inline src/main/java/net/minecraftforge/coremod/api/ASMAPI.java:53]: I've marked this method as deprecated for removal, but should it be removed? It's existed in CoreMods for a long time and a lot of older CoreMods, some of which might still be used today in FMLOnly-type mods, might be using this method.

File: src/main/java/net/minecraftforge/coremod/api/ASMAPI.java [hunk 2]
```diff
@@ -361,35 +387,42 @@ public static MethodInsnNode findFirstMethodCallBefore(MethodNode method, Method
...
+
+    /**
+     * Inserts/replaces a list after/before the given instruction.
+     *
+     * @param method The method where you want to insert the list
+     * @param list   The list that should be inserted
+     * @param mode   How the given code should be inserted
+     * @return True if the list was inserted, false otherwise
+     */
+    public static boolean insertInsnList(MethodNode method, AbstractInsnNode insn, InsnList list, InsertMode mode) {
```
Comment (Jonathing) [inline src/main/java/net/minecraftforge/coremod/api/ASMAPI.java:414]: This method returns `boolean` to have parity with the older `insertInsnList` method. However, I think that it might be better to crash the game if a modder tries to insert a list into the instructions using an instruction that is not included in the method. What do you think?

Comment (LexManos) [reply src/main/java/net/minecraftforge/coremod/api/ASMAPI.java:414]: Having it return a boolena means the CoreMod itself can determine to crash or not

File: src/main/java/net/minecraftforge/coremod/api/ASMAPI.java [hunk 3]
```diff
@@ -80,17 +91,17 @@ public static MethodInsnNode buildMethodCall(final String ownerName, final Strin
     }
 
     /**
-     * Signifies the type of number constant for a {@link LdcNumberType}.
+     * Signifies the type of number constant for a {@link NumberType}.
      */
-    public enum LdcNumberType {
+    public enum NumberType {
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/coremod/api/ASMAPI.java:96]: Why was this renamed/does it break compatibility? How old is this?

Comment (Jonathing) [reply src/main/java/net/minecraftforge/coremod/api/ASMAPI.java:96]: It was renamed because it is no longer exclusive to `LdcInsnNode`. There aren't any versions of Forge that have bumped CoreMods recently to include `LdcNumberType`, so it doesn't break compatibility.

Comment (LexManos) [reply src/main/java/net/minecraftforge/coremod/api/ASMAPI.java:96]: Okay, then if you're done i can pull this

Comment (Jonathing) [reply src/main/java/net/minecraftforge/coremod/api/ASMAPI.java:96]: Yeah it's ready
